### ISSUE TITLE
[AssistedInject] Support multiple methods in factories

### ISSUE THREAD

**Issue by yankeppey** [google/dagger#2354]
First of all, I'm really happy to see that Dagger finally supports assisted injection, thank you so much!

Comparing  to [AutoFactory](https://github.com/google/auto/tree/master/factory), it makes developer to write more boilerplate code, but gives more freedom in how to define an interface. However, it seems an unnecessary limitation to be forced to have _just one_  method in a class annotated as an @AssistedFactory.

In a large project I'm trying to integrate Dagger into there are many factories like this:

```
class FruitFactory(...) {
   fun createApple(someDep: SomeDep, someDep2: SomeDep2, dependencyApple: DependencyApple...): Apple {
     //...
   }
   fun createOrange(someDep3: SomeDep3, dependencyOrange: DependencyOrange...): Orange {
     //...
   }
   fun createPear(dependencyPear: DependencyPear...): Pear {
     //...
   }
}
```

In the end, I'd really like to declare it as the following and make Dagger to do the rest:

```
@AssistedFactory
interface FruitFactory {
   fun createApple(dependencyApple: DependencyApple): Apple

   fun createOrange(dependencyOrange: DependencyOrange): Orange

   fun createPear(dependencyPear: DependencyPear): Pear
}
```

However, with the current state of Dagger (2.31.2), I have to declare three distinct classes of PearFactory, AppleFactory and OrangeFactory, and then either manually wrap them into FruitFactory (seems wasteful) or inject them separately, forgetting about FruitFactory (which may be a reasonable abstraction).

Are there plans to remove this limitation in future?

**Comment by bcorso** [comment]
Yeah, I think that sounds reasonable. We can look into supporting this.

**Comment by bentatham** [comment]
I agree - this would be really useful. I recently upgraded an application from dagger1 to dagger2. We were/are using AutoFactory, and there are a few cases where there are multiple factory methods created for us for the multiple constructors. I was hoping to get rid of AutoFactory while upgrading to dagger2 as well, because why not use the built-in support, but can't do that (at least not across the board) until this is done.

**Comment by danlieb-stripe** [comment]
+1 from me. Any chance this will happen any time soon @bcorso?

**Comment by bcorso** [comment]
> My use case is perhaps slightly different. I'd like to be able to support multiple different factories for the same type, but with different parameters with overloaded AssistedInject constructors.

@danlieb-stripe, I don't think we're going to support multiple `@AssistedInject` constructors; however, you should be able to get something similar using the assisted factory, e.g.

```java
public class Foo {
  @AssistedInject
  Foo(@Assisted Bar bar, @Assisted Baz baz) { ... }
}

@AssistedFactory
public abstract class Factory {
  // This is implemented by Dagger
  abstract Foo create(Bar bar, Baz baz);

  // Factory method 1
  public Foo create(Bar bar) {
    return create(bar, DEFAULT_BAZ_VALUE);
  }

  // Factory method 2
  public Foo create(Baz baz) {
    return create(DEFAULT_BAR_VALUE, baz);
  }
}
```
